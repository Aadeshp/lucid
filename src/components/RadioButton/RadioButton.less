@RadioButton-height: 22px;
@RadioButton-width: 22px;
@RadioButton-animation-timing: 0.1s linear;

@RadioButton-container-offset: 3px;
@RadioButton-container-radius: 8px;

@RadioButton-dot-offset: 8px;
@RadioButton-dot-radius: 3px;

@RadioButton-glow-offset: 3px;
@RadioButton-glow-radius: 8px;
@RadioButton-glow-scale: scale(1.375);

.lucid-RadioButton {
	cursor: pointer;
	display: inline-block;
	height: @RadioButton-height;
	position: relative;
	user-select: none;
	width: @RadioButton-width;

	&-is-disabled {
		cursor: default;
	}

	&-native {
		opacity: 0;
	}

	&-visualization-container {
		background-color: white;
		border-radius: @RadioButton-container-radius + 1;
		border: solid 1px rgba(0, 0, 0, 0.3);
		height: @RadioButton-container-radius * 2;
		left: @RadioButton-container-offset;
		position: absolute;
		top: @RadioButton-container-offset;
		width: @RadioButton-container-radius * 2;
	}

	&-visualization-dot {
		background-color: white;
		border-radius: @RadioButton-dot-radius;
		height: @RadioButton-dot-radius * 2;
		left: @RadioButton-dot-offset;
		opacity: 0;
		position: absolute;
		top: @RadioButton-dot-offset;
		transition: @RadioButton-animation-timing;
		width: @RadioButton-dot-radius * 2;
	}

	&-visualization-glow {
		background-color: @color-primary;
		border-radius: @RadioButton-glow-radius;
		height: @RadioButton-glow-radius * 2;
		left: @RadioButton-glow-offset;
		opacity: 0;
		position: absolute;
		top: @RadioButton-glow-offset;
		transition: @RadioButton-animation-timing;
		width: @RadioButton-glow-radius * 2;
	}

	&:hover &-visualization-container {
		border-color: @color-primary;
		transition: @RadioButton-animation-timing;
	}

	&:hover &-visualization-glow {
		opacity: 0.3;
		transform: @RadioButton-glow-scale;
	}

	&-is-disabled &-visualization-container,
	&-is-disabled:hover &-visualization-container {
		background-color: rgba(0, 0, 0, 0.05);
		border-color: rgba(0, 0, 0, 0.1);
	}

	&-is-disabled:hover &-visualization-glow {
		opacity: 0;
		transform: none;
	}

	&-is-disabled&-is-selected &-visualization-container {
		opacity: 0.3;
	}

	&-is-disabled&-is-selected:hover &-visualization-container,
	&-is-selected &-visualization-container {
		background-color: @color-primary;
		border-color: @color-primary;
	}

	&-is-selected &-visualization-dot {
		opacity: 1;
	}

	&-native:focus ~ &-visualization-glow {
		opacity: 0.3;
		transform: @RadioButton-glow-scale;
	}
}
