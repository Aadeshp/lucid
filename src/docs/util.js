import _ from 'lodash';
import { markdown } from 'markdown';

export function toMarkdown(md) {
	if (md) {
		return {
			__html: markdown.toHTML(md),
		};
	}

	return {
		__html: '',
	};
}

export function handleHighlightCode() {
	if (window.hljs) {
		//eslint-disable-line
		_.each(document.querySelectorAll('pre code'), block => {
			hljs.highlightBlock(block); //eslint-disable-line
		});
	}
}

export function sanitizeExamplePath(path) {
	return _.flow(
		x => _.split(x, '/'),
		x => _.compact(x),
		x => _.drop(x),
		x => _.join(x, '-'),
		x => _.split(x, '.'),
		x => _.join(x, '-')
	)(path);
}

export function stripIndent(text) {
	if (!text) {
		return '';
	}

	const sigLines = text.split(/\n/).filter(line => !/^\s*$/.test(line));
	const leadingWhitespaces = sigLines.map(line => /^(\s*)/g.exec(line)[1]);
	const minimumLeadingWhitespace = leadingWhitespaces.reduce(
		(min, line) => (line.length < min.length ? line : min),
		leadingWhitespaces[0]
	);
	return text
		.split(/\n/)
		.map(line => line.replace(minimumLeadingWhitespace, ''))
		.join('\n');
}
