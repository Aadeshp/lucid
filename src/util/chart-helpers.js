import _ from 'lodash';
import d3TimeFormat from 'd3-time-format';

/**
 * groupByFields
 *
 * This will return the data in a similar format to d3Shape.stack
 * but without the stacking of the data.
 *
 * @param {object[]} collection - normalized data you want to operate on
 * @param {string[]} fields - fields to pluck off for the y data
 * @return {array[]} - array of arrays, one for each field
 */
export function groupByFields(collection, fields) {
	const fieldsArray = [].concat(fields);

	return _.map(fieldsArray, (field) => {
		return _.map(collection, (item) => item[field]);
	});
}

/**
 * toPairs
 *
 * Takes a collection of data and pairs up some x and y data.
 *
 * @param {object[]} collection
 * @param {string} xField
 * @param {string} yField
 * @return {array} - E.g. [[x, y], [x, y]]
 */
export function toPairs(collection, xField, yField) {
	return _.map(collection, (item) => [item[xField], item[yField]]);
}

/**
 * byFields
 *
 * Takes a collection of data and returns an array of all the fields off that
 * collection.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {array}
 */
export function byFields(collection, fields) {
	const fieldArray = [].concat(fields);

	return _.reduce(fieldArray, (acc, field) => {
		return acc.concat(_.map(collection, field));
	}, []);
}

/**
 * minByFields
 *
 * Returns the minimum element from a collection by a set of fields.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function minByFields(collection, fields) {
	return _.min(byFields(collection, fields));
}

/**
 * maxByFields
 *
 * Returns the maximum element from a collection by a set of fields.
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function maxByFields(collection, fields) {
	return _.max(byFields(collection, fields));
}

/**
 * maxByFieldsStacked
 *
 * Returns the max sum of a set of fields from a collection
 *
 * @param {object[]} collection
 * @param {string[]} fields
 * @return {any}
 */
export function maxByFieldsStacked(collection, fields) {
	const fieldArray = [].concat(fields);

	const sums = _.reduce(collection, (acc, item) => {
		const sum = _.chain(item)
			.pick(fieldArray)
			.toArray()
			.sum()
			.value();
		return acc.concat(sum);
	}, []);

	return _.max(sums);
}

/**
 * discreteTicks
 *
 * Returns `count` evenly spaced, representative values from the `array`.
 *
 * @param {array} array
 * @param {number} size - should be greater than 1
 * @return {array}
 */
export function discreteTicks(array, count) {
	if (!array || _.isNil(count) || array.length <= count) {
		return array;
	}

	const step = (array.length - 1) / Math.max(1, count - 1);

	return _.reduce(_.times(count), (acc, n) => {
		return acc.concat(array[Math.round(n  * step)]);
	}, []);
}

/**
 * transformFromCenter
 *
 * Scaling paths from their center is tricky. This function
 * helps do that be generating a translate/scale transform
 * string with the correct numbers.
 *
 * @param {number} x - the x data point where you want the path to be centered at
 * @param {number} y - the y data point where you want the path to be centered at
 * @param {number} xCenter - the x coordinate of the center of the path you're trying to transform
 * @param {number} yCenter - the x coordinate of the center of the path you're trying to transform
 * @param {number} scale - number to scale to, 2 would be 2x bigger
 * @return {string} - transform string
 */
export function transformFromCenter(x, y, xCenter, yCenter, scale) {
	return `translate(${((1 - scale) * xCenter) + (x - xCenter)}, ${((1 - scale) * yCenter) + (y - yCenter)}) scale(${scale})`;
}

const formatMillisecond = d3TimeFormat.timeFormat('.%L');
const formatSecond = d3TimeFormat.timeFormat(':%S');
const formatMinute = d3TimeFormat.timeFormat('%I:%M');
const formatHour = d3TimeFormat.timeFormat('%I %p');
const formatDay = d3TimeFormat.timeFormat('%a %d');
const formatWeek = d3TimeFormat.timeFormat('%b %d');
const formatMonth = d3TimeFormat.timeFormat('%B');
const formatYear = d3TimeFormat.timeFormat('%Y');

export function multiFormat(date) {
	return (d3TimeFormat.timeSecond(date) < date ? formatMillisecond
		: d3TimeFormat.timeMinute(date) < date ? formatSecond
		: d3TimeFormat.timeHour(date) < date ? formatMinute
		: d3TimeFormat.timeDay(date) < date ? formatHour
		: d3TimeFormat.timeMonth(date) < date ? (d3TimeFormat.timeWeek(date) < date ? formatDay : formatWeek)
		: d3TimeFormat.timeYear(date) < date ? formatMonth
		: formatYear)(date);
}

