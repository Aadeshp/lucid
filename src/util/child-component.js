import React from 'react';
import _ from 'lodash';

/**
 * createChildComponent
 *
 * Returns a React Component which does not render anything, but has some useful static functions
 *
 * @return {ReactComponent} which renders `null` but exposes static helper functions
 */
export function createChildComponent (definition={}) {
	return React.createClass({
		propTypes: _.get(definition, 'propTypes', null),
		displayName: _.get(definition, 'displayName', null),
		render: () => null,
		statics: {
			isRenderNull: true,

			/**
			 * findInChildren
			 *
			 * Returns all elements with this component type
			 *
			 * @param {array} elements - usually `this.props.children` but can be any list of elements which might have this type
			 * @return {array} elements with this component type
			 */
			findInChildren(elements) {
				let component = this;
				let ownElements = [];
				React.Children.forEach(elements, (element) => {
					if (element && element.type === component) {
						ownElements.push(element);
					}
				});
				return ownElements;
			},

			/**
			 * findInChildrenAsProps
			 *
			 * Returns the `props` object of each element with this component type
			 *
			 * @param {array} elements - usually `this.props.children` but can be any list of elements which might have this type
			 * @return {array} `props` objects of elements with this component type
			 */
			findInChildrenAsProps(elements) {
				let component = this;
				return _.map(component.findInChildren(elements), 'props');
			},

			/**
			 * findInProps
			 *
			 * Given a set of parent props this will return only the props for this
			 * component type
			 *
			 * @param {object} parentProps - usually `this.props` that may contain a prop for the current component type
			 * @return {array} - an array of normalized props
			 */
			findInProps(parentProps) {
				// Extract out the defined `childProps` from the parent class
				const currentProps = _.chain(parentProps)
					.get(definition.propName) // grab the prop we care about, e.g. "Child"
					.thru(x => [x]) // wrap in an array
					.flatten()
					.filter(x => x) // remove falsey values
					.value();

				return _.map(currentProps, (childProp) => {
					if (_.isPlainObject(childProp)) {
						return childProp;
					} else {
						return {
							children: childProp
						};
					}
				});
			},

			/**
			 * findInAllAsProps
			 *
			 * Given a parent props object, this will return a mashup of all the
			 * children and/or props for the current component type. This is
			 * particularly useful for providing a flexible component API. It allows
			 * consumers to either pass stuff through props or children.
			 *
			 * @param {object} parentProps - usually `this.props` that may contain a prop for the current component type
			 * @return {array} - an array of `props`
			 */
			findInAllAsProps(parentProps) {
				let component = this;

				const propsFromProps = component.findInProps(parentProps);
				const propsFromChildren = component.findInChildrenAsProps(parentProps.children);

				return propsFromProps.concat(propsFromChildren);
			}
		}
	});
}

export function rejectNullElements(children) {
	return _.reject(React.Children.toArray(children), (node) => (React.isValidElement(node) && node.type.isRenderNull));
}
