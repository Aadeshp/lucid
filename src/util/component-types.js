import React from 'react';
import _ from 'lodash';

// creates a React component
export function createClass(definition={}) {
	const {
		_isPrivate = false,
		statics = {},
		components = {},
		reducers = {},
		selectors = {},
		propName = null,
		propTypes = {},
		render = () => null,
		...restDefinition
	} = definition;

	const newDefinition = {
		...restDefinition,
		statics: {
			...statics,
			...components,
			_isPrivate,
			reducers,
			selectors,
			propName,
		},
		propTypes: _.assign({}, propTypes, _.mapValues(definition.components, () => React.PropTypes.any)),
		render,
	};

	newDefinition.statics.definition = newDefinition;

	return React.createClass(newDefinition);
}

// return all elements matching the specified types
export function filterTypes(children, types=[]) {
	types = [].concat(types); // coerce to Array

	return _.filter(
		React.Children.toArray(children),
		(element) => (React.isValidElement(element) && _.includes(types, element.type))
	);
}

// return all elements not matching the specified types
export function rejectTypes(children, types=[]) {
	types = [].concat(types); // coerce to Array

	return _.reject(
		React.Children.toArray(children),
		(element) => (React.isValidElement(element) && _.includes(types, element.type))
	);
}

// return an array of elements (of the given type) for each of the values
export function createElements(type, values=[]) {
	return _.reduce([].concat(values), (elements, typeValue) => {
		if (React.isValidElement(typeValue) && typeValue.type === type) {
			return elements.concat(typeValue);
		} else if (_.isPlainObject(typeValue) && !React.isValidElement(typeValue)) {
			return elements.concat(React.createElement(type, typeValue));
		} else if (_.isUndefined(typeValue)) {
			return elements;
		} else {
			return elements.concat(React.createElement(type, null, typeValue));
		}
	}, []);
}

// return all elements found in props and children of the specified types
export function findTypes(props, types=[]) {
	// get elements from props (using type.propName)
	const elementsFromProps = _.reduce(_.castArray(types), (acc, type) => {
		if (!_.isUndefined(type.propName)) {
			const propMatches = _.flatten(_.values(_.pick(props, type.propName)));
			return acc.concat(createElements(type, propMatches));
		}
		return acc;
	}, []);

	// return elements from props and elements from children
	return elementsFromProps.concat(filterTypes(props.children, types));
}

// return the first element found in props and children of the specificed type(s)
export function getFirst(props, types, defaultValue) {
	return _.first(findTypes(props, types)) || defaultValue;
}

// omit props defined in propTypes of the given type
// and any extra keys given in third argument
export function omitProps(props, type, keys = []) {
	return _.omit(props, _.keys(type.propTypes).concat(keys).concat(['initialState']));
}
