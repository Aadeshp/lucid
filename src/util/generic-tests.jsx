import sinon from 'sinon';
import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import describeWithDOM  from './describe-with-dom';
import _ from 'lodash';
import * as lucid from '../index';

// Common tests for all our components
export function common(Component, getDefaultProps=_.noop) {

	function generateDefaultProps(props={}) {
		return _.assign({}, getDefaultProps(), props);
	}

	describe(`[common]`, () => {
		it('should have a `displayName` defined', () => {
			assert(Component.displayName);
		});

		it('should pass through styles to the root element', () => {
			let style = {
				backgroundColor: '#f0f'
			};
			const wrapper = shallow(<Component {...generateDefaultProps()} style={style}/>);
			assert.deepEqual(wrapper.first().prop('style'), style);
		});

		it('should pass through `className`', () => {
			let expectedClass = 'rAnDoM';
			const wrapper = shallow(<Component {...generateDefaultProps()} className={expectedClass}/>);
			let classNames = wrapper.first().prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${expectedClass}'`);
		});

		it('should have an application scoped base class', () => {
			let expectedClass = 'lucid-' + Component.displayName;
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			let classNames = wrapper.first().prop('className').split(' ');

			assert(_.includes(classNames, expectedClass), `'${classNames}' should include '${Component.displayName}'`);
		});

		it('should have only application scoped classes', () => {
			const wrapper = shallow(<Component {...generateDefaultProps()} />);
			let parentClasses = wrapper.first().prop('className').split(' ');
			let childrenClasses = wrapper.children().reduce((acc, node) => {
				if (!node.prop('className')) {
					return acc;
				}

				return acc.concat(node.prop('className').split(' '))
			}, []);

			let allClasses = parentClasses.concat(childrenClasses);

			assert(_.every(allClasses, (className) => {
				return _.includes(className, 'lucid-' + Component.displayName);
			}));
		});

		it('should only use onX convention for function proptypes', () => {
			assert(_.every(Component.propTypes, (value, key) => {
				// See the following tests from the React source code to figure out how
				// this works: https://github.com/facebook/react/blob/v0.14.7/src/isomorphic/classic/types/__tests__/ReactPropTypes-test.js
				let props = {};
				props[key] = _.noop;

				// Here we test the validation against a noop function, if it doesn't
				// have an error, then it's probably a function. The other possible is
				// that it's an `any`, so we'll test for that below
				let isProbablyFunction = !(value(props, key) instanceof Error);

				// We'll also test out a string to see if that valid
				props[key] = '';

				let isAny = isProbablyFunction && !(value(props, key) instanceof Error);

				// If it's probably a function, and it's not `any`, then we make sure
				// it starts with `on`
				if (isProbablyFunction && !isAny && !_.startsWith(key, 'on')) {
					return false
				}

				return true;
			}));
		});

		it('should be available as an exported module from index.js', () => {
			assert(lucid[Component.displayName]);
		});
	});
}

// Common tests for all our icon components
export function icons(Component) {
	describeWithDOM(`[icon]`, () => {
		it('should pass through isBadge prop to underlying Icon component', () => {
			const wrapper = mount(<Component isBadge={true} />);
			const classNames = wrapper.find('svg').prop('className').split(' ');
			const targetClassName = 'lucid-Icon-is-badge';
			assert(_.includes(classNames, targetClassName), `'${classNames}' should include '${targetClassName}'`);
		});
	});
}

// Common tests for all control components
export function controls(Component, { callbackName, controlSelector , eventType }) {
	// Use DOM tests here since some of our controls use dom events under the hood
	describeWithDOM('[control]', () => {
		it('should callback with `uniqueId` and `event` prop', () => {
			const expectedUniqueId = 32;
			const props = {
				uniqueId: expectedUniqueId,
				[callbackName]: sinon.spy(),
			};
			const wrapper = mount(<Component {...props} />);

			wrapper.find(controlSelector).simulate(eventType);

			// Last argument should be an object with `uniqueId` and `event`
			const { uniqueId, event } = _.last(props[callbackName].args[0]);

			assert(event, 'missing event');
			assert.equal(uniqueId, expectedUniqueId, 'incorrect or missing uniqueId');
		});
	});
}
