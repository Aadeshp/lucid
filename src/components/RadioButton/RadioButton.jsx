import _ from 'lodash';
import classNames from 'classnames';
import React from 'react';

import { bindClassNames } from '../../util/style-helpers';

const generateBoundClassNames = bindClassNames('RadioButton');
const {
	bool,
	func,
	object,
	string
} = React.PropTypes;

/**
 * {"categories": ["controls", "toggles"]}
 *
 * This is a toggle -- a component that is in one of two particular states at
 * any given moment in time -- that features a custom visualization of the
 * native radio button control to reflect its current state.
 *
 * The radio button is different from a standard toggle in that when it is in
 * the selected state user events do not cause it to toggle to the unselected
 * state so the `onSelect` function is called only when `isSelected` is false.
 *
 * It uses a hidden native radio button control under the hood but leverages
 * other HTML elements to visualize its state.
 */
const RadioButton = React.createClass({
	propTypes: {
		/**
		 * Appended to the component-specific class names set on the root
		 * element.
		 */
		className: string,

		/**
		 * Indicates whether the component should appear and act disabled by
		 * having a "greyed out" palette and ignoring user interactions.
		 */
		isDisabled: bool,

		/**
		 * Indicates that the component is in the "selected" state when true
		 * and in the "unselected" state when false.
		 */
		isSelected: bool,

		/**
		 * Called when the user clicks on the component or when they press the
		 * space key while the component is in focus, and only called when the
		 * component is in the unselected state. The first parameter is true,
		 * and the second parameter is the synthetic event object generated by
		 * React.
		 */
		onSelect: func,

		/**
		 * Passed through to the root element.
		 */
		style: object
	},

	getDefaultProps() {
		return {
			isDisabled: false,
			isSelected: false,
			onSelect: _.noop
		};
	},

	componentDidMount() {
		this.nativeElement = this.refs.nativeElement;
	},

	render() {
		const {
			className,
			isDisabled,
			isSelected,
			style,
			...passThroughs
		} = this.props;

		return (
			<div
					className={classNames(generateBoundClassNames('~', {
						'is-disabled': isDisabled,
						'is-selected': isSelected
					}), className)}
					onClick={this.handleClicked}
					onTouchEnd={this.handleClicked}
					style={style}
			>
				<input
						{...passThroughs}
						checked={isSelected}
						className={generateBoundClassNames('native')}
						disabled={isDisabled}
						defaultChecked={isSelected}
						ref='nativeElement'
						type='radio'
				/>
				<span className={generateBoundClassNames('visualization-glow')} />
				<span className={generateBoundClassNames('visualization-container')} />
				<span className={generateBoundClassNames('visualization-dot')} />
			</div>
		);
	},

	handleClicked(event) {
		const {
			isDisabled,
			isSelected,
			onSelect
		} = this.props;

		if (!isDisabled && !isSelected) {
			onSelect(true, event);
			this.nativeElement.focus();
		}
	}
});

export default RadioButton;
