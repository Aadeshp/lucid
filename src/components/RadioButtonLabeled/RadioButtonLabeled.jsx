import _ from 'lodash';
import React from 'react';
import { lucidClassNames } from '../../util/style-helpers';
import { createClass, findTypes }  from '../../util/component-types';
import RadioButton from '../RadioButton/RadioButton';

const cx = lucidClassNames.bind('&-RadioButtonLabeled');
const {
	any,
	node,
	object,
	string,
} = React.PropTypes;

/**
 * {"categories": ["controls", "toggles"], "extend": "RadioButton", "madeFrom": ["RadioButton"]}
 *
 * This is a composite of the `RadioButton` component and the native `label`
 * element.
 */
const RadioButtonLabeled = createClass({
	displayName: 'RadioButtonLabeled',

	components: {
		Label: createClass({
			displayName: 'RadioButtonLabeled.Label',
			propName: 'Label',
			propTypes: {
				/**
				 * Used to identify the purpose of this radio button to the user --
				 * can be any renderable content.
				 */
				children: node,
			},
		}),
	},

	propTypes: {
		...RadioButton.propTypes,

		/**
		 * Appended to the component-specific class names set on the root
		 * element.
		 */
		className: string,

		/**
		 * Passed through to the root element.
		 */
		style: object,

		/**
		 * Child element whose children are used to identify the purpose of this
		 * radio button to the user.
		 */
		Label: any,
	},

	getDefaultProps() {
		return {
			isDisabled: false,
			isSelected: false,
			onSelect: _.noop,
		};
	},

	render() {
		const {
			className,
			isDisabled,
			isSelected,
			onSelect,
			style,
			...passThroughs,
		} = this.props;

		const labelChildProps = _.first(_.map(findTypes(this.props, RadioButtonLabeled.Label), 'props'));

		return (
			<label
					className={cx('&', {
						'&-is-disabled': isDisabled,
						'&-is-selected': isSelected,
					}, className)}
					style={style}
			>
				<RadioButton
						className={className}
						isDisabled={isDisabled}
						isSelected={isSelected}
						onSelect={onSelect}
						{..._.omit(passThroughs, 'Label')}
				/>
				{
					labelChildProps
							? labelChildProps.children
							: null
				}
			</label>
		);
	},
});

export default RadioButtonLabeled;
