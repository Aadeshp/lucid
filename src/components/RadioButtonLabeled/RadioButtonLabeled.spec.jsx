import _ from 'lodash';
import assert from 'assert';
import React from 'react';
import SyntheticEvent from 'react/lib/SyntheticEvent';
import sinon from 'sinon';
import { mount, shallow } from 'enzyme';

import describeWithDOM from '../../util/describe-with-dom';
import { common } from '../../util/generic-tests';

import RadioButtonLabeled from './RadioButtonLabeled';
import RadioButton from '../RadioButton/RadioButton';

SyntheticEvent;
sinon;
describeWithDOM;

describe('RadioButtonLabeled', () => {
	common(RadioButtonLabeled);

	describe('props', () => {
		describe('isDisabled', () => {
			it('passes the value through to its `RadioButton` instance.', () => {
				const wrapper = shallow(<RadioButtonLabeled isDisabled={true} />);

				assert.equal(wrapper.find(RadioButton).prop('isDisabled'), true);
			});

			it('defaults to `false`.', () => {
				const wrapper = mount(<RadioButtonLabeled />);

				assert.equal(wrapper.prop('isDisabled'), false);
			});
		});

		describe('isSelected', () => {
			it('passes the value through to its `RadioButton` instance.', () => {
				const wrapper = shallow(<RadioButtonLabeled isSelected={true} />);

				assert.equal(wrapper.find(RadioButton).prop('isSelected'), true);
			});

			it('defaults to `false`.', () => {
				const wrapper = mount(<RadioButtonLabeled />);

				assert.equal(wrapper.prop('isSelected'), false);
			});
		});

		describe('onSelect', () => {
			it('passes the value through to its `RadioButton` instance.', () => {
				const foo = () => null;
				const wrapper = shallow(<RadioButtonLabeled onSelect={foo} />);

				assert.equal(wrapper.find(RadioButton).prop('onSelect'), foo);
			});

			it('defaults to the Lodash `noop` method.', () => {
				const wrapper = mount(<RadioButtonLabeled />);

				assert.equal(wrapper.prop('onSelect'), _.noop);
			});
		});

		describe('pass throughs', () => {
			it('passes through all props not defined in `propTypes` to its `RadioButton` instance.', () => {
				const wrapper = shallow(
					<RadioButtonLabeled
							className='wut'
							isDisabled={true}
							isSelected={true}
							style={{ fontWeight: 'bold' }}
							onSelect={_.noop}
							foo={1}
							bar={2}
							baz={3}
							qux={4}
							quux={5}
					/>
				);
				const radioButtonProps = wrapper.find(RadioButton).props();

				// It should pass `foo`, `bar`, `baz`, `qux`, and `quux` through
				// to the `RadioButton` instance.
				_.forEach(['foo', 'bar', 'baz', 'qux', 'quux'], (prop) => {
					assert(_.has(radioButtonProps, prop));
				});
			});
		});
	});
});
