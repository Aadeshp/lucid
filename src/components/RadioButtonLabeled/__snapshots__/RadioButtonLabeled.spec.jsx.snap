exports[`RadioButtonLabeled [common] example testing should match snapshot for interactive 1`] = `
<section>
  <span>
    <RadioButtonLabeled
      isDisabled={false}
      isSelected={true}
      name="interactive-radio-buttons"
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
        }
      }>
      <RadioButtonLabeled.Label>
        Vanilla
      </RadioButtonLabeled.Label>
    </RadioButtonLabeled>
    <RadioButtonLabeled
      isDisabled={false}
      isSelected={false}
      name="interactive-radio-buttons"
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
        }
      }>
      <RadioButtonLabeled.Label>
        Chocolate
      </RadioButtonLabeled.Label>
    </RadioButtonLabeled>
    <RadioButtonLabeled
      isDisabled={false}
      isSelected={false}
      name="interactive-radio-buttons"
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
        }
      }>
      <RadioButtonLabeled.Label>
        Strawberry
      </RadioButtonLabeled.Label>
    </RadioButtonLabeled>
    <RadioButtonLabeled
      isDisabled={false}
      isSelected={false}
      name="interactive-radio-buttons"
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
        }
      }>
      <RadioButtonLabeled.Label>
        Polaroid four dollar toast bespoke succulents. Kickstarter truffaut PBR&B fashion axe, lyft actually viral everyday carry iPhone tote bag mumblecore umami. Skateboard you probably haven\'t heard of them before they sold out, edison bulb paleo poutine jianbing blue bottle mixtape. Normcore farm-to-table coloring book cliche before they sold out. Roof party authentic hoodie paleo next level, bicycle rights selvage you probably haven\'t heard of them leggings venmo etsy cronut williamsburg. Dreamcatcher vice gastropub austin fam. Actually subway tile kickstarter, messenger bag shabby chic activated charcoal lomo.
      </RadioButtonLabeled.Label>
    </RadioButtonLabeled>
    <RadioButtonLabeled
      isDisabled={false}
      isSelected={false}
      name="interactive-radio-buttons"
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
        }
      }>
      <RadioButtonLabeled.Label>
        Unicorn helvetica glossier pop-up letterpress snackwave raw denim. Lomo blog tattooed helvetica seitan DIY. Tbh kombucha shabby chic cornhole, microdosing seitan snackwave gastropub celiac. Biodiesel 90\'s fixie drinking vinegar literally unicorn readymade fap, vape wolf direct trade kickstarter pickled disrupt fashion axe. Heirloom subway tile hell of, austin portland meditation PBR&B tacos raclette. Prism fanny pack banh mi trust fund unicorn, thundercats chicharrones salvia pour-over VHS. Distillery organic pork belly squid.
      </RadioButtonLabeled.Label>
    </RadioButtonLabeled>
    <RadioButtonLabeled
      isDisabled={false}
      isSelected={false}
      name="interactive-radio-buttons"
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
        }
      }>
      <RadioButtonLabeled.Label>
        Heirloom keffiyeh schlitz cray. Keytar tumblr dreamcatcher art party polaroid, sriracha etsy pinterest typewriter. IPhone ugh williamsburg man bun waistcoat literally paleo direct trade. Taxidermy cold-pressed pabst green juice, freegan tumblr post-ironic everyday carry disrupt distillery. Hell of lumbersexual cliche, whatever tilde raclette pop-up blue bottle skateboard butcher meditation four dollar toast tattooed health goth pinterest. Poutine polaroid YOLO, biodiesel chambray gentrify microdosing kickstarter bushwick humblebrag yr cred intelligentsia you probably haven\'t heard of them. Blog meditation locavore tousled, edison bulb cronut occupy +1 you probably haven\'t heard of them echo park distillery chambray.
      </RadioButtonLabeled.Label>
    </RadioButtonLabeled>
  </span>
</section>
`;

exports[`RadioButtonLabeled [common] example testing should match snapshot for label-as-child 1`] = `
<section>
  <section>
    <RadioButtonLabeled
      isDisabled={false}
      isSelected={false}
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
        }
      }>
      <RadioButtonLabeled.Label>
        Just text
      </RadioButtonLabeled.Label>
    </RadioButtonLabeled>
    <RadioButtonLabeled
      isDisabled={false}
      isSelected={false}
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
        }
      }>
      <RadioButtonLabeled.Label>
        <span>
          HTML element
        </span>
      </RadioButtonLabeled.Label>
    </RadioButtonLabeled>
  </section>
</section>
`;

exports[`RadioButtonLabeled [common] example testing should match snapshot for label-as-prop 1`] = `
<section>
  <section>
    <RadioButtonLabeled
      Label="Just text"
      isDisabled={false}
      isSelected={false}
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
        }
      } />
    <RadioButtonLabeled
      Label={
        <span>
          HTML element
        </span>
      }
      isDisabled={false}
      isSelected={false}
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
        }
      } />
    <RadioButtonLabeled
      Label={
        Array [
          "Text in an array",
          "Only the first value in the array is used",
          "The rest of these should be ignored",
        ]
      }
      isDisabled={false}
      isSelected={false}
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
        }
      } />
    <RadioButtonLabeled
      Label={
        Array [
          <span>
            HTML element in an array
        </span>,
          <span>
            Again only the first value in the array is used
        </span>,
          <span>
            The rest should not be rendered
        </span>,
        ]
      }
      isDisabled={false}
      isSelected={false}
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
        }
      } />
  </section>
</section>
`;

exports[`RadioButtonLabeled [common] example testing should match snapshot for states 1`] = `
<section>
  <RadioButtonLabeled
    isDisabled={false}
    isSelected={false}
    onSelect={[Function]}
    style={
      Object {
        "marginBottom": "3px",
        "marginRight": "13px",
      }
    }>
    <RadioButtonLabeled.Label>
      (default props)
    </RadioButtonLabeled.Label>
  </RadioButtonLabeled>
  <section
    style={
      Object {
        "display": "flex",
      }
    }>
    <RadioButtonLabeled
      isDisabled={true}
      isSelected={false}
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
          "marginRight": "13px",
        }
      }>
      <RadioButtonLabeled.Label>
        Disabled
      </RadioButtonLabeled.Label>
    </RadioButtonLabeled>
    <RadioButtonLabeled
      isDisabled={false}
      isSelected={true}
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
          "marginRight": "13px",
        }
      }>
      <RadioButtonLabeled.Label>
        Selected
      </RadioButtonLabeled.Label>
    </RadioButtonLabeled>
    <RadioButtonLabeled
      isDisabled={true}
      isSelected={true}
      onSelect={[Function]}
      style={
        Object {
          "marginBottom": "3px",
          "marginRight": "13px",
        }
      }>
      <RadioButtonLabeled.Label>
        Disabled & selected
      </RadioButtonLabeled.Label>
    </RadioButtonLabeled>
  </section>
</section>
`;
